File: python/1283.py
class Solution:
    def smallestDivisor(self, nums: List[int], threshold: int) -> int:
        begin, end = 1, max(nums)+1
        while begin < end:
            mid = (end - begin)//2 + begin
            _sum = 0
            for num in nums:
                _sum += (num + mid - 1) // mid
            if _sum <= threshold:
                end = mid
            else:
                begin = mid + 1
        return end

File: cpp/1283.cpp
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    int smallestDivisor(vector<int>& nums, int threshold) {
        int end = *max_element(nums.cbegin(), nums.cend());
        int begin = 1;
        int mid, _sum, res = -1;
        while (end > begin) {
            mid = (end - begin)/2 + begin;
            _sum = 0;
            for (int num: nums) {
                _sum += (num - 1 + mid)/mid;
            }

            if (_sum > threshold) {
                begin = mid+1;
            } else {
                end = mid;
            }
        }
        return end;

    }


};